### PR TITLE
Add `download-xsd` CLI command

### PR DESCRIPTION
**Pull request by leebradley**
Add an XSD download CLI command, along with several utility methods for
downloading, extracting, and validating the files within the schema
archive.

Signed-off-by: Lee Bradley <bradley@bitwise.io>

### CODE REVIEW

**Review by dnewh — DISMISSED** [review]
tested

File: cli/src/actions/xsd_downloader/validator.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+use crate::error::CliError;
+
+/// Validate the hash of a file an return a useful error message if the hash is invalid
+///
+/// * `file` - The file to check the hash of
+/// * `hash` - The expected sha256 hash
+pub fn validate_hash(file: &Path, hash: &[u8; 32]) -> Result<(), CliError> {
+    info!("validating hash of {file}", file = file.to_string_lossy());
+    let mut hasher = Sha256::new();
+    let data = std::fs::read(file).unwrap();
```
Comment (CLEckhardt) [inline cli/src/actions/xsd_downloader/validator.rs:15]: --no-download panics when there isn't an artifact

```
root@gridd-alpha:/usr/share/grid/xsd# grid download-xsd --no-downl│
oad                                                               │
validating hash of /var/lib/grid/artifact/gs1.zip                 │
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` v│
alue: Os { code: 2, kind: NotFound, message: "No such file or dire│
ctory" }', cli/src/actions/xsd_downloader/validator.rs:15:36      │
note: run with `RUST_BACKTRACE=1` environment variable to display │
a backtrace      
```

File: cli/src/main.rs [hunk 2]
```diff
@@ -1775,6 +1777,50 @@ fn run() -> Result<(), CliError> {
...
+                        .conflicts_with("copy_from")
+                        .help("Never download a file, only extract the cached file"),
+                )
+                .arg(
+                    Arg::with_name("no_checksum")
+                        .long("no-checksum")
+                        .help("Do not perform a checksum on the cached file"),
+                )
+                .arg(
+                    Arg::with_name("copy_from")
```
Comment (CLEckhardt) [inline cli/src/main.rs:1810]: It looks like this should take a value.

File: cli/src/main.rs [hunk 3]
```diff
@@ -1775,6 +1777,50 @@ fn run() -> Result<(), CliError> {
...
+                        .conflicts_with("force_download")
+                        .conflicts_with("no_download")
+                        .long("copy-from")
+                        .help("Replenish the cache from a directory resource and use that"),
+                )
+                .arg(
+                    Arg::with_name("artifact_dir")
```
Comment (CLEckhardt) [inline cli/src/main.rs:1817]: It looks like this should also take a value.

**Review by CLEckhardt — CHANGES_REQUESTED** [review]

File: cli/src/actions/xsd_downloader/validator.rs (hunk 1 above)
Comment (leebradley) [reply cli/src/actions/xsd_downloader/validator.rs]: Good catch! Fixed

File: cli/src/main.rs (hunk 2 above)
Comment (leebradley) [reply cli/src/main.rs:1797]: Added `takes_value(true)`

File: cli/src/main.rs (hunk 3 above)
Comment (leebradley) [reply cli/src/main.rs]: Added `takes_value(true)`

File: cli/man/grid-download-xsd.1.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
+% GRID-KEYGEN(1) Cargill, Incorporated | Grid
+
+<!--
+  Copyright 2021 Cargill Incorporated
```
Comment (shannynalayna) [inline cli/man/grid-download-xsd.1.md:4]: This man page should be copyright for `2022`

File: cli/src/actions/xsd_downloader/downloader.rs [hunk 5]
```diff
@@ -0,0 +1,419 @@
+use std::fs::{self, File};
```
Comment (shannynalayna) [inline cli/src/actions/xsd_downloader/downloader.rs:1]: This file is missing a copyright

File: cli/src/actions/xsd_downloader/extractor.rs [hunk 6]
```diff
@@ -0,0 +1,264 @@
+use std::fs::{self, File};
```
Comment (shannynalayna) [inline cli/src/actions/xsd_downloader/extractor.rs:1]: Need a copyright

File: cli/src/actions/xsd_downloader/mod.rs [hunk 7]
```diff
@@ -0,0 +1,603 @@
+mod downloader;
```
Comment (shannynalayna) [inline cli/src/actions/xsd_downloader/mod.rs:1]: Copyright :)

Comment (shannynalayna) [reply cli/src/actions/xsd_downloader/mod.rs:1]: Copyrights need to be added to all the new files, noticed a few more but won't comment everywhere

File: cli/src/main.rs [hunk 8]
```diff
@@ -1775,6 +1774,54 @@ fn run() -> Result<(), CliError> {
...
+                .arg(
+                    Arg::with_name("force")
+                        .long("force")
+                        .help("Continue even if a checksum on the cached file is incorrect"),
+                )
+                .arg(
+                    Arg::with_name("copy_from")
+                        .takes_value(true)
+                        .conflicts_with("force_download")
+                        .conflicts_with("no_download")
```
Comment (shannynalayna) [inline cli/src/main.rs:1800]: From the man page, it looks like this option should be usable with `no_download`. The man page should be updated to show these args are not usable together, unless this line is a mistake?

File: cli/man/grid-download-xsd.1.md (hunk 4 above)
Comment (leebradley) [reply cli/man/grid-download-xsd.1.md]: Updated!

File: cli/src/actions/xsd_downloader/downloader.rs (hunk 5 above)
Comment (leebradley) [reply cli/src/actions/xsd_downloader/downloader.rs:17]: Added!

File: cli/src/actions/xsd_downloader/extractor.rs (hunk 6 above)
Comment (leebradley) [reply cli/src/actions/xsd_downloader/extractor.rs:17]: Added!

File: cli/src/actions/xsd_downloader/mod.rs (hunk 7 above)
Comment (leebradley) [reply cli/src/actions/xsd_downloader/mod.rs:17]: Added!

File: cli/src/main.rs (hunk 8 above)
Comment (leebradley) [reply cli/src/main.rs]: So actually I'd somehow missed the req that no_download should be compatible with copy_from, so you found a legit bug here. Fixed and added a unit test to validate the functionality.

**Review by shannynalayna — DISMISSED** [review]
Tested

**Review by CLEckhardt — COMMENTED** [review]
Tested and functionality is good. A few notes:
- `--copy-from` without `--no-download` effectively ignores the arg if the filename is incorrect
- `--force` still gives a message “validating hash of ...”
- man page and CLI help should probably say that the file for `--copy-from` needs to be named “gs1.zip”
Still working on testing po actions, running into some unrelated issues, but I'll confirm it all works when that's done

**Comment by leebradley** [comment]
> * `--copy-from` without `--no-download` effectively ignores the arg if the filename is incorrect

Yeah, I was initially going to be very strict and require all files exist, but there's also the use case @shannynalayna pointed out of only having some of the files available (and downloading the rest). I'm not sure how to best approach this.

> * `--force` still gives a message “validating hash of ...”

By design `--force` still validates the hash of downloaded files, and only ignores the hashes of cached files.

> * man page and CLI help should probably say that the file for `--copy-from` needs to be named “gs1.zip”

Hmm if this is going to be added to docs that means we'll technically be supporting manually tinkering with the internal cache, which tbh I didn't anticipate. I think I'm going to rename this file so it's a little more future-proof.

**Comment by CLEckhardt** [comment]
Cool, makes sense.

**Comment by leebradley** [comment]
So in response to the feedback form @CLEckhardt I've made three changes:

1) After consulting a bit, we've decided to support users working with these cache files. Thus I've renamed "gs1.zip" in the cache to "GS1_XML_3-4-1_Publication.zip" so it matches the default filename and will not need renamed in the future if a different version comes out.

2) Added a reference to GS1_XML_3-4-1_Publication.zip in the docs so users are aware what needs to be in the copy directory.

3) Changed a `debug!` message to `info!` so users can immediately see if there are missing files in the cache directory. Example below:

```
root@gridd-alpha:/# grid download-xsd --copy-from foo              
file "GS1_XML_3-4-1_Publication.zip" doesn't exist in copy path
downloading file from https://www.gs1.org/docs/EDI/xml/3.4.1/GS1_XML_3-4-1_Publication.zip
download finished
validating hash of /var/cache/grid/xsd_artifact_cache/GS1_XML_3-4-1_Publication.zip.download
extracting to schema directory
```

File: cli/man/grid-download-xsd.1.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+**grid download-xsd** \[**FLAGS**\] \[**OPTIONS**\] 
+
+DESCRIPTION
+===========
+
+This command downloads GS1 XSD files used by various Grid features. The
+downloaded artifacts are first copied into a cache directory. They are then
+expanded into Grid's state directory. If the desired artifacts are in the
+cahce directory, Grid will not attempt to re-download them, and instead
```
Comment (shannynalayna) [inline cli/man/grid-download-xsd.1.md:26]: sp/cahce/cache

Comment (leebradley) [reply cli/man/grid-download-xsd.1.md]: Fixed!

**Review by shannynalayna — APPROVED** [review]
Tested; temp file is being removed if the overall download fails

**Review by dnewh — APPROVED** [review]

**Pull request merged**